plugins/experimental/cookie_remap/cookie_remap.cc

```diff
@@ -79,17 +79,24 @@ struct UrlComponents {
   }
 
   void
-  populate(TSRemapRequestInfo *rri)
+  populate(TSRemapRequestInfo *rri, TSHttpTxn txn)
   {
-    scheme    = TSUrlSchemeGet(rri->requestBufp, rri->requestUrl, &scheme_len);
-    host      = TSUrlHostGet(rri->requestBufp, rri->requestUrl, &host_len);
-    tspath    = TSUrlPathGet(rri->requestBufp, rri->requestUrl, &tspath_len);
-    query     = TSUrlHttpQueryGet(rri->requestBufp, rri->requestUrl, &query_len);
-    matrix    = TSUrlHttpParamsGet(rri->requestBufp, rri->requestUrl, &matrix_len);
-    port      = TSUrlPortGet(rri->requestBufp, rri->requestUrl);
-    url       = TSUrlStringGet(rri->requestBufp, rri->requestUrl, &url_len);
+    TSMBuffer bufp;
+    TSMLoc urlp;
+    if (TSHttpTxnPristineUrlGet(txn, &bufp, &urlp) != TS_SUCCESS) {
+      return;
+    }
+    scheme    = TSUrlSchemeGet(bufp, urlp, &scheme_len);
+    host      = TSUrlHostGet(bufp, urlp, &host_len);
+    tspath    = TSUrlPathGet(bufp, urlp, &tspath_len);
+    query     = TSUrlHttpQueryGet(bufp, urlp, &query_len);
+    matrix    = TSUrlHttpParamsGet(bufp, urlp, &matrix_len);
+    port      = TSUrlPortGet(bufp, urlp);
+    url       = TSUrlStringGet(bufp, urlp, &url_len);
     from_path = TSUrlPathGet(rri->requestBufp, rri->mapFromUrl, &from_path_len);
 
+    TSHandleMLocRelease(bufp, TS_NULL_MLOC, urlp);
```

Comment:
This seems odd to me. `TSHttpTxnPristineUrlGet()` returns a marshal buffer in `bufp` and location handle in `urlp`. Then `TSUrlSchemeGet()`, `TSUrlHostGet()`, etc. are called and the pointers pointing **inside** the marshal buffer are saved and used later **after** `TSHandleMLocRelease()` is called.

Although we don't do/use any reference counting currently and TSHandleMLocRelease() is a no-op for this type of objects (`HDR_HEAP_OBJ_URL`) I think we should make sure we don't release before we are done with the internal pointers.

Depending on the plugin needs we could (1) store the values pointed by the internal pointers and use them after the release or (2) keep the `bufp` and `urlp` until we are done with the pointers (for the lifetime of the `UrlComponents` class), similarly to cachekey plugin ([class CacheKey](https://github.com/apache/trafficserver/blob/master/plugins/cachekey/cachekey.cc#L249)) or tscppapi ([class ClientRequest](https://github.com/apache/trafficserver/blob/master/src/tscpp/api/ClientRequest.cc#L47)).

Reply:
Good catch. Will fix it.

Reply:
I would be _very_ cautious about keeping the raw internal pointers, as those can go stale during any header access call. It is the `TSMLoc` that's stable, AFAICT that's the whole point of that type.

Reply:
Good point! In fact this is what the cachekey and tscppapi do in the examples provided above.